pynumdiff/kalman_smooth/_kalman_smooth.py

```diff
@@ -6,638 +6,255 @@
-
-    if y is not None and not np.isnan(y):
-        xhat_fp = xhat_fm + K_f@(y - C@xhat_fm)
-        P_fp = (I - K_f@C)@P_fm
-        xhat_fm = A@xhat_fp + B@u
-        P_fm = A@P_fp@A.T + gammaW@Q@gammaW.T
-    else:
-        xhat_fp = xhat_fm
-        P_fp = (I - K_f@C)@P_fm
-        xhat_fm = A@xhat_fp + B@u
-        P_fm = A@P_fp@A.T + gammaW@Q@gammaW.T
-        
-    return xhat_fp, xhat_fm, P_fp, P_fm
-
-
-def __kalman_forward_filter__(xhat_fm, P_fm, y, u, A, B, C, R, Q):
-    """
-    :param xhat_fm:
-    :param P_fm:
-    :param y:
-    :param u:
-    :param A:
-    :param B:
-    :param C:
-    :param R:
-    :param Q:
-    :return:
-    """
-    if u is None:
-        u = np.array(np.zeros([B.shape[1], y.shape[1]]))
-
-    xhat_fp = None
-    P_fp = []
-    P_fm = [P_fm]
-
-    for i in range(y.shape[1]):
-        _xhat_fp, _xhat_fm, _P_fp, _P_fm = __kalman_forward_update__(xhat_fm[:, [-1]], P_fm[-1], y[:, [i]], u[:, [i]],
-                                                                     A, B, C, R, Q)
-        if xhat_fp is None:
-            xhat_fp = _xhat_fp
-        else:
-            xhat_fp = np.hstack((xhat_fp, _xhat_fp))
-        xhat_fm = np.hstack((xhat_fm, _xhat_fm))
-
-        P_fp.append(_P_fp)
-        P_fm.append(_P_fm)
-
-    return xhat_fp, xhat_fm, P_fp, P_fm
-
-
-def __kalman_backward_smooth__(xhat_fp, xhat_fm, P_fp, P_fm, A):
+def _kalman_forward_filter(xhat0, P0, y, A, C, Q, R, u=None, B=None):
+    """Run the forward pass of a Kalman filter
+    :param np.array xhat0: initial estimate of state (often 0) one step before the first measurement
+    :param np.array P0: initial guess of state covariance (often identity matrix) before the first measurement
+    :param np.array y: series of measurements, stacked along axis 0. In this case just the raw noisy data (fully observable)
+    :param np.array A: discrete-time state transition matrix
+    :param np.array C: measurement matrix
+    :param np.array Q: discrete-time process noise covariance
+    :param np.array R: measurement noise covariance
+    :param np.array u: optional control input
+    :param np.array B: optional control matrix
+    :return: tuple[np.array, np.array, np.array, np.array] of\n
+        - **xhat_pre** -- a priori estimates of xhat, with axis=0 the batch dimension, so xhat[n] gets the nth step
+        - **xhat_post** -- a posteriori estimates of xhat
+        - **P_pre** -- a priori estimates of P
+        - **P_post** -- a posteriori estimates of P
     """
-    :param xhat_fp:
-    :param xhat_fm:
-    :param P_fp:
-    :param P_fm:
-    :param A:
-    :return:
-    """
-    N = xhat_fp.shape[1]
-
-    xhat_smooth = copy.copy(xhat_fp)
-    P_smooth = copy.copy(P_fp)
-    for t in range(N-2, -1, -1):
-        L = P_fp[t]@A.T@np.linalg.pinv(P_fm[t])
-        xhat_smooth[:, [t]] = xhat_fp[:, [t]] + L@(xhat_smooth[:, [t+1]] - xhat_fm[:, [t+1]])
-        P_smooth[t] = P_fp[t] - L@(P_smooth[t+1] - P_fm[t+1])
```

Comment:
But this line is incorrect, because you're supposed to use $P_{t+1|t}$ here as well, which is stored in `P_fm[t]`. https://arc.aiaa.org/doi/10.2514/3.3166

This line is also wrong, because you're subtracting the `L` part instead of adding it. So this isn't doing RTS smoothing at all. Serious bug.